src/Microsoft.Performance.SDK/Processing/ProcessingSource.cs

```diff
@@ -0,0 +1,391 @@
+        ///     found via reflection.
+        /// </param>
+        /// <param name="tableAssemblyProvider">
+        ///     A function that provides the <see cref="Assembly"/> that should be searched
+        ///     for tables.
+        /// </param>
+        /// <exception cref="System.ArgumentNullException">
+        ///     <paramref name="additionalTablesProvider"/> is <c>null</c>.
+        ///     - or -
+        ///     <paramref name="tableAssemblyProvider"/> is <c>null</c>.
+        /// </exception>
+        protected ProcessingSource(
+            Func<IDictionary<TableDescriptor, Action<ITableBuilder, IDataExtensionRetrieval>>> additionalTablesProvider,
+            Func<Assembly> tableAssemblyProvider)
+        {
+            Guard.NotNull(additionalTablesProvider, nameof(additionalTablesProvider));
+            Guard.NotNull(tableAssemblyProvider, nameof(tableAssemblyProvider));
+
+            this.allTables = new Dictionary<TableDescriptor, Action<ITableBuilder, IDataExtensionRetrieval>>();
+            this.allTablesRO = new ReadOnlyDictionary<TableDescriptor, Action<ITableBuilder, IDataExtensionRetrieval>>(this.allTables);
+            this.metadataTables = new HashSet<TableDescriptor>();
+            this.metadataTablesRO = new ReadOnlyHashSet<TableDescriptor>(this.metadataTables);
+
+            this.tableAssemblyProvider = tableAssemblyProvider;
+            this.additionalTablesProvider = additionalTablesProvider;
+        }
+
+        /// <inheritdoc />
+        public IEnumerable<TableDescriptor> DataTables => this.allTables.Keys
+            .Except(this.metadataTablesRO);
+
+        /// <inheritdoc />
+        public IEnumerable<TableDescriptor> MetadataTables => this.metadataTablesRO;
+
+        /// <inheritdoc />
+        public virtual IEnumerable<Option> CommandLineOptions => Enumerable.Empty<Option>();
+
+        /// <summary>
+        ///     Gets a mapping of <see cref="TableDescriptor"/> to the concrete
+        ///     <see cref="Type" /> of Table described by the descriptor. This
+        ///     mapping includes the data and metadata tables.
+        /// </summary>
+        protected IReadOnlyDictionary<TableDescriptor, Action<ITableBuilder, IDataExtensionRetrieval>> AllTables => this.allTablesRO;
+
+        /// <summary>
+        ///     This is set by default when the SetApplicationEnvironment is called by the runtime.
+        ///     <seealso cref="IApplicationEnvironment"/>
+        /// </summary>
+        protected IApplicationEnvironment ApplicationEnvironment { get; private set; }
+
+        /// <summary>
+        ///     This is set when <see cref="SetLogger"/> is called by the runtime.
+        ///     <seealso cref="ILogger"/>
+        /// </summary>
+        protected ILogger Logger { get; private set; }
+
+        /// <inheritdoc />
+        /// <summary>
+        ///     When overridden in a derived class, saves the given environment into this instance.
```

Comment:
>When overridden in a derived class, saves the given environment into this instance. [](http://example.com/codeflow?start=16&length=83)

We should update this comment. Derived classes should override SetApplicationEnvironmentCore instead.

Reply:
This original comment doesn't make sense to me either. If this method is getting overridden, how is it saving the given application environment?